astropy/io/fits/hdu/table.py

```diff
@@ -885,22 +885,21 @@ def _calculate_datasum_with_heap(self):
             # Now add in the heap data to the checksum (we can skip any gap
             # between the table and the heap since it's all zeros and doesn't
             # contribute to the checksum
-            # TODO: The following code may no longer be necessary since it is
-            # now possible to get a pointer directly to the heap data as a
-            # whole.  That said, it is possible for the heap section to contain
-            # data that is not actually pointed to by the table (i.e. garbage;
-            # this *shouldn't* happen but it is not disallowed either)--need to
-            # double check whether or not the checksum should include such
-            # garbage
```

Comment:
I went and read the [FITS  4](https://fits.gsfc.nasa.gov/standard40/fits_standard40aa-le.pdf) standard which includes the compression and checksum extensions. The wording for DATASUM is:

> The value field of the DATASUM keyword shall consist of a character string that should contain the unsigned-integer value of the 32-bit ones’ complement checksum of the data records in the HDU (i.e., excluding the header records). For this purpose, each 2880-byte FITS logical record should be interpreted as consisting of 720 32-bit unsigned integers.

which reads to me like the whole FITS block for all the blocks that are present should be included irrespective of what they are used for?

Reply:
And this would be consistent with what is done with reading a file I guess ?

Reply:
For calculating the checksum? yes.